File: Kalorie/Kalorie/Core/UseCases/CreateMealTypeUseCase.swift
//
//  CreateMealTypeUseCase.swift
//  Kalorie
//
//  Created by Josef Antoni on 29.06.2026.
//

import Foundation
import MealKit

protocol CreateMealTypeUseCaseProtocol {
    func callAsFunction(
        name: String,
        startTime: Date,
        endTime: Date,
        existingMealTypes: [MealTypeDomain]
    ) async throws -> MealTypeDomain
}

struct CreateMealTypeUseCase: CreateMealTypeUseCaseProtocol {

    // MARK: - Properties

    private let dataProvider: any FirestoreDataProviderProtocol
    private let authProvider: any AuthProviderProtocol

    // MARK: - Init

    init(dataProvider: any FirestoreDataProviderProtocol, authProvider: any AuthProviderProtocol) {
        self.dataProvider = dataProvider
        self.authProvider = authProvider
    }

    // MARK: - Functions

    func callAsFunction(
        name: String,
        startTime: Date,
        endTime: Date,
        existingMealTypes: [MealTypeDomain]
    ) async throws -> MealTypeDomain {
        guard !name.isEmpty else { throw CreateMealTypeError.emptyName }
        guard !existingMealTypes.contains(where: { $0.name == name }) else {
            throw CreateMealTypeError.duplicateName
        }
        let startMinutes = startTime.minutesSinceMidnight
        let endMinutes = endTime.minutesSinceMidnight
        guard MealWindowsKt.isMealWindowLongEnough(startMinutes: startMinutes, endMinutes: endMinutes, minimumDurationMinutes: 30) else {
            throw CreateMealTypeError.durationTooShort
        }
        guard !existingMealTypes.contains(where: {
            MealWindowsKt.mealWindowsOverlap(
                startMinutes: startMinutes,
                endMinutes: endMinutes,
                otherStartMinutes: $0.startTime.minutesSinceMidnight,
                otherEndMinutes: $0.endTime.minutesSinceMidnight
            )
        }) else {
            throw CreateMealTypeError.timeConflict
        }
        guard let userId = authProvider.userId else { throw AuthError.notAuthenticated }
        let newId = UUID().uuidString
        let dto = MealTypeDTO(
            id: newId,
            name: name,
            startMinutes: Int(startMinutes),
            endMinutes: Int(endMinutes)
        )
        try await dataProvider.setAsync(dto, id: newId, in: Constants.Firestore.mealTypes(userId: userId))
        return MealTypeDomain(id: newId, name: name, startTime: startTime, endTime: endTime)
    }
}

#if DEBUG
struct CreateMealTypeUseCaseFake: CreateMealTypeUseCaseProtocol {

    // MARK: - Functions

    func callAsFunction(
        name: String,
        startTime: Date,
        endTime: Date,
        existingMealTypes: [MealTypeDomain]
    ) async throws -> MealTypeDomain {
        MealTypeDomain(id: UUID().uuidString, name: name, startTime: startTime, endTime: endTime)
    }
}
#endif

File: Android/app/src/main/java/antoni/kalorie/core/usecases/CreateMealTypeUseCase.kt
package antoni.kalorie.core.usecases

import antoni.kalorie.core.auth.AuthError
import antoni.kalorie.core.auth.AuthProviderProtocol
import antoni.kalorie.core.models.MealTypeDomain
import antoni.kalorie.core.networking.FirestoreDataProviderProtocol
import antoni.kalorie.core.networking.MealTypeDTO
import antoni.kalorie.core.networking.setAsync
import antoni.kalorie.core.utils.Constants
import antoni.kalorie.features.mealtypesheet.CreateMealTypeError
import antoni.kalorie.mealkit.MIN_MEAL_WINDOW_MINUTES
import antoni.kalorie.mealkit.isMealWindowLongEnough
import antoni.kalorie.mealkit.mealWindowsOverlap
import java.util.UUID

interface CreateMealTypeUseCaseProtocol {
    suspend operator fun invoke(
        name: String,
        startMinutes: Int,
        endMinutes: Int,
        existingMealTypes: List<MealTypeDomain>,
    ): MealTypeDomain
}

class CreateMealTypeUseCase(
    private val dataProvider: FirestoreDataProviderProtocol,
    private val authProvider: AuthProviderProtocol,
) : CreateMealTypeUseCaseProtocol {

    // MARK: - Functions

    override suspend fun invoke(
        name: String,
        startMinutes: Int,
        endMinutes: Int,
        existingMealTypes: List<MealTypeDomain>,
    ): MealTypeDomain {
        val name = name.trim()
        if (name.isEmpty()) throw CreateMealTypeError.EmptyName
        val comparableName = name.lowercase()
        if (existingMealTypes.any { it.name.trim().lowercase() == comparableName }) {
            throw CreateMealTypeError.DuplicateName
        }
        if (!isMealWindowLongEnough(startMinutes, endMinutes, MIN_MEAL_WINDOW_MINUTES)) {
            throw CreateMealTypeError.DurationTooShort
        }
        if (existingMealTypes.any { mealWindowsOverlap(startMinutes, endMinutes, it.startMinutes, it.endMinutes) }) {
            throw CreateMealTypeError.TimeConflict
        }
        val userId = authProvider.userId ?: throw AuthError.NotAuthenticated
        val newId = UUID.randomUUID().toString().uppercase()
        val dto = MealTypeDTO(id = newId, name = name, startMinutes = startMinutes, endMinutes = endMinutes)
        dataProvider.setAsync(dto, id = newId, inCollection = Constants.Firestore.mealTypes(userId))
        return MealTypeDomain(id = newId, name = name, startMinutes = startMinutes, endMinutes = endMinutes)
    }
}
